rcl/src/rcl/timer.c

```diff
@@ -294,15 +294,15 @@ rcl_timer_call(rcl_timer_t * timer)
   // between the timer being ready and the callback being triggered
   next_call_time += period;
   // in case the timer has missed at least once cycle
-  if (next_call_time < now) {
+  if (next_call_time <= now) {
```

Comment:
what if the period of one case? technically we expect executor calls timer on each tick as much as possible, this will lead the much less cyclic to call the timer?

Reply:
> next_call_time will be now + 1ns

i think this is wrong, it should be `next_call_time = next_call_time + period`.

and adding if statement `(next_call_time == now)`, means either we should call the timer callback or we already miss the timer at least one cycle.
i think the former is correct, timer should be called with `(next_call_time == now)` case. So `periods_ahead` should not be added.

Reply:
> i think this is wrong, it should be next_call_time = next_call_time + period.
I was just referring to your example of `period = 1`, that would result in a `next_call_time` equal to `now + 1`, because `next_call_time` will get increased by a multiple of `period` until its greater than `now`.

next_call_time will be set to `next_call_time += period` in line 295. But if the result is still in the past (or equal to now with my change) it will get increased by a multiple of `period` so that the result will not be in the past.

> and adding if statement (next_call_time == now), means either we should call the timer callback or we already miss the timer at least one cycle.
We are currently in the timer call, the callback will get called in line 313. If `next_call_time == now` we missed the timer already, but since we are about to call it the next call should definitely be in the future. If not, the timer will get called with the same timestemp again, in case of using sim_time. (Because sim_time is updated rarely compared to system clock)

Reply:
> We are currently in the timer call, the callback will get called in line 313.

i don't think so. according to your sample, you are using `rclcpp`, right? so that this should be NULL.

user callback will be called `rclcpp` here,

https://github.com/ros2/rclcpp/blob/fbe8f28cd13710c5c643a4e7149e509f3a952677/rclcpp/include/rclcpp/timer.hpp#L279

> If next_call_time == now we missed the timer already,

for me, this means that we should not set the bias. instead, we go to the next execution right away to call the timer to avoid missing the timer cyclic as less as possible. and then period increases `next_call_time`.

Reply:
> i don't think so. according to your sample, you are using rclcpp, right? so that this should be NULL.

True, rclcpp does the callback differently, I just wanted to stay in scope of `rcl`. Anyway the decision for calling is based on `next_call_time` in both implementations and `rcl_timer_call` is called in both cases to update the timer internals.

> for me, this means that we should not set the bias. instead, we go to the next execution right away to call the timer to avoid missing the timer cyclic as less as possible. and then period increases next_call_time.

To prevent the issue from occurring `next_call_time` must be different from the current call time, which is `now`.

I think the current implementation tries to stick to the period between cycles, even if previous cycles were missed. That's why `next_call_time` is always incremented by a multiple of `period` instead of just setting it to `now + 1` to minimise missed cycles.

But I am fine with any solution, as long as multiple callback calls with the same timestamp are prevented.

Reply:
How can we procced? Should we @mention someone to ask for a second opinion?

Reply:
@kohrt yes, that i would like to do. actually i had a quick chat with @ros2/team we will get the feedback, please give us some time.

Reply:
I spent a bit of time looking at this, and I have a question and some comments.

First, I want to ask about this statement:

> If next_call_time == now we missed the timer already,

Is that really true?  If we look at the implementation of `rcl_timer_is_ready` (the next function down), we determine that the timer is ready if the (next_call_time - now **<=** 0).  So it seems to me that the timer is considered ready (and thus expired) in both the case where the next call time is prior to now, and when they are exactly equal.  Does that match your understanding?

If what I say above is true, then I think it makes sense for us to change this check to **<=** as well, since `next_call_time == now` is considered "ready".  We could consider changing `rcl_timer_is_ready` to change that **<=** to a **<**, but I think that would cause us to miss more timer periods.

Getting slightly more formal here, we should consider what happens in this bit of code for all three cases of `next_call_time < now`, `next_call_time == now`, and `next_call_time > now`.  The third case is illogical; we should never get here if that is the case.  Nevertheless, it might be a good idea to check for that in a separate clause just to cover all of the bases.  The first case is already covered by the current code.  That leaves the equal case, and like I said it seems to make sense to me to make this change because we consider **<=** to be ready below.

Reply:
> First, I want to ask about this statement:
> 
> > If next_call_time == now we missed the timer already,
> 
> Is that really true? If we look at the implementation of `rcl_timer_is_ready` (the next function down), we determine that the timer is ready if the (next_call_time - now **<=** 0). So it seems to me that the timer is considered ready (and thus expired) in both the case where the next call time is prior to now, and when they are exactly equal. Does that match your understanding?

Yes, that is true.

But maybe we are talking about different things. Let me try to clarify the issue:

We are at timestamp `now` and the current timer had been checked for expiration, `rcl_timer_is_ready` was called and set `is_ready = true`, therefor `rcl_timer_call` is called.

That means that we are in the state that the timer is expired and the timer in currently in the process of being called.

One part of calling the timer is to update the timestamps for the next timer call.

`next_call_time` will be increased by the timer period: `next_call_time += period;`

Afterwards it is checked, if the `next_call_time` is still in the past: `if (next_call_time < now)`.
*Note: If it's the case, it means that the timer should have been called at least period time earlier, but the period was missed.*

If `next_call_time` is in the past, it is increased by a multiple of period so that the result is `>= now`.

This can lead to the situation that the current call time is `now` and `next_call_time` is also `now`. In case the system time does not update as fast as the timer is checked for expiration, the timer will to be called twice with the same timestamp. This can be the case when sim_time is used.

Therefore I suggested to change the update of  `next_call_time` to always give results that are in the future. That's why I changed the check to `<=` and increased update by +1 period.

Currently `next_call_time` is always increased by a multiple of period, even if a previous period was missed. I think this is done on purpose, therefore my suggestion keeps it that way.

Alternatively, one could set `next_call_time = now + 1;` in case of a missed period, to minimise the overall missed periods.

*Note: The special case of a timer with a period of 0 is not considered yet. But if we change it, to always set the 'next_call_time' to a future timestamps, we should consider changing the special case update in line 300 to `next_call_time = now + 1;` as well.*

Reply:
@clalancette thanks for taking a look at this.

I am good to go with you and @kohrt , i just wanted to have the 2nd opinion since me and @kohrt were only having this discussion 😄 appreciate it.

@kohrt thanks for being patient on this 👍